flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/utils/PlannerMock.java

```diff
@@ -50,12 +50,12 @@ public String explain(List<Operation> operations, ExplainDetail... extraDetails)
     }
 
     @Override
-    public CompiledPlan loadPlan(PlanReference planReference) throws IOException {
+    public CompiledPlanInternalFactory loadPlan(PlanReference planReference) throws IOException {
```

Comment:
I don't like this additional indirection. The long `CompiledPlanInternalFactory` name already shows how "not so nice" the current approach is compared to the simplicity of the current `Planner` interface.

How about we simply throw an `UnsupportedOperationException` in the `CompiledPlan.execute()` returned by the planner by default and the TableEnvironment can simply wrap the `CompiledPlan` one more time to provide the executable functionality.

Another option would be to pass `TableEnvironmentInternal` to `loadPlan` and `compilePlan` but this is also not very nice because it creates a cyclic dependency.

Reply:
TBH I think this is the way it should look, as this factory just underlines how I need a dependency injection before being able to use an object fulfilling the `CompiledPlan` interface. The fact that the name is not nice is a different problem, we can rename it to `CompiledPlanFactory` if you prefer.

IMHO using `UnsupportedOperationException` is not correct, as the object returned by `loadPlan` here would not respect the interface `CompiledPlan`, which is `Executable` and `Explainable`. It also makes reasoning about this code and the functionalities of these interfaces harder, as now I need to go look into the specific `ExecNodeCompiledPlan` implementation to find out that I need an additional wrapping to be able to implement the full `CompiledPlan` interface. I rather prefer to make `ExecNodeCompiledPlan` mutable and clarify that I need additional dependency injection before being able to use it.

Reply:
The `Explainable` was working before. We are just discussing `Executable` here. I don't think it makes a difference whether we throw an `UnsupportedOperationException` for `execute()` or throw an exception if the dependency injection has not performed.

Passing a `Function<CompiledPlan, TableResult>` argument is also not nice.

I don't have a better idea. So let's just shorten the name to `CompiledPlanFactory` and stick to this design. But I have the feeling the class structure gets more and more complex for no reason. Also the need for `CompiledPlanInternal` is weak. `TableEnvironmentInternal` has actually a comment about `Once old planner is removed, this class also can be removed.`

Reply:
I pushed the name change.

One other possible solution is to make `CompiledPlanInternal` not extend `CompiledPlan`, and add to `CompiledPlanInternal` the `writeToFile`, `asJsonString` and `getFlinkVersion`. This way, the "user functionalities" of `CompiledPlan` are implemented in a wrapper class, which implements the full contract of `CompiledPlan`. Then `ExecNodeCompiledPlan` needs to implement only these 3 methods + `getSinkIdentifiers`, and the rest of the "user fluent APIs" are implemented by the wrapper.

This IMO really solves the issue in an elegant way, although a bit verbosely, but only internally in the `TableEnvironmentImpl`.

Reply:
I think properly splitting API from internal `CompiledPlan` makes sense. I was also thinking about this.

So:
- `CompiledPlan` = API
- `InternalPlan` = used for all `Planner` methods

Let's try to not use `CompiledPlanInternal` if it doesn't extend from `CompiledPlan` anymore.